File: sdk-python/rtmemory/conversations.py
"""ConversationsNamespace — async methods for the /v1/conversations/ API."""

from __future__ import annotations

from typing import Any

import httpx

from rtmemory.types import (
    ConversationAddRequest,
    ConversationAddResponse,
    ConversationMessage,
)


class ConversationsNamespace:
    """Namespace for conversation memory operations."""

    def __init__(self, http: httpx.AsyncClient) -> None:
        self._http = http

    async def add(
        self,
        messages: list[dict[str, str] | ConversationMessage],
        space_id: str,
        user_id: str | None = None,
    ) -> ConversationAddResponse:
        """Submit a conversation fragment (triggers extraction)."""
        parsed_messages = []
        for m in messages:
            if isinstance(m, ConversationMessage):
                parsed_messages.append(m)
            else:
                parsed_messages.append(ConversationMessage(**m))
        body = ConversationAddRequest(messages=parsed_messages, space_id=space_id, user_id=user_id)
        resp = await self._http.post("/v1/conversations/", json=body.model_dump(exclude_none=True))
        resp.raise_for_status()
        return ConversationAddResponse.model_validate(resp.json())

    async def end(
        self,
        conversation_id: str,
        space_id: str,
    ) -> dict[str, Any]:
        """End a conversation (triggers deep scan)."""
        body = {"conversation_id": conversation_id, "space_id": space_id}
        resp = await self._http.post("/v1/conversations/end", json=body)
        resp.raise_for_status()
        return resp.json()

File: sdk-js/src/conversations.ts
/**
 * ConversationClient — thin wrapper around the /v1/conversations/ API.
 */

import type {
  ConversationAddRequest,
  ConversationAddResponse,
} from "./types";
import { RTMemoryError } from "./client";

export class ConversationClient {
  private baseUrl: string;
  private headers: Record<string, string>;

  constructor(baseUrl: string, headers: Record<string, string>) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.headers = { ...headers };
  }

  /** Submit a conversation fragment (triggers extraction). */
  async add(req: ConversationAddRequest): Promise<ConversationAddResponse> {
    const res = await fetch(`${this.baseUrl}/v1/conversations/`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(req),
    });
    if (!res.ok) throw new RTMemoryError(res.status, await res.text());
    return res.json() as Promise<ConversationAddResponse>;
  }

  /** End a conversation (triggers deep scan). */
  async end(conversationId: string, spaceId: string): Promise<ConversationAddResponse> {
    const res = await fetch(`${this.baseUrl}/v1/conversations/end`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify({ conversation_id: conversationId, space_id: spaceId }),
    });
    if (!res.ok) throw new RTMemoryError(res.status, await res.text());
    return res.json() as Promise<ConversationAddResponse>;
  }
}
